File: server/Services/VaultKeepsService.cs
using Microsoft.AspNetCore.Authentication.OpenIdConnect;
using Microsoft.Extensions.Configuration.UserSecrets;

namespace Keepr.Services;

public class VaultKeepsService
{
    private readonly VaultKeepsRepository _repository;
    private readonly VaultsService _vaultsService;
    private readonly KeepsService _keepsService;

    public VaultKeepsService(VaultKeepsRepository repository, VaultsService vaultsService, KeepsService keepsService)
    {
        _repository = repository;
        _vaultsService = vaultsService;
        _keepsService = keepsService;
    }

    internal KeepClone CreateVaultKeep(VaultKeep vaultKeepData)
    {
        // FIXME get vault first to see if I am the creator of it (CRATE REPORT ON RESTAURANT REVIEWS)
        Vault vault = _vaultsService.GetVaultById(vaultKeepData.VaultId, vaultKeepData.CreatorId);

        if (vault.CreatorId != vaultKeepData.CreatorId)
        {
            throw new Exception("You don't have permission!");
        }

        int keepId = vaultKeepData.KeepId;
        _keepsService.KeptKeep(keepId);
        KeepClone vaultKeep = _repository.CreateVaultKeep(vaultKeepData);
        //ANCHOR KeepClone vaultKeep 
        // VaultKeep vaultKeep = _repository.CreateVaultKeep(vaultKeepData);
        return vaultKeep;
    }

    internal List<KeepClone> GetVaultKeepsByVaultId(int vaultId, string userId)
    {
        _vaultsService.GetVaultById(vaultId, userId);
        List<KeepClone> keepClones = _repository.GetVaultKeepsByVaultId(vaultId);
        return keepClones;
    }

    internal KeepClone GetVaultKeepById(int vaultKeepId)
    {
        KeepClone keepClone = _repository.GetVaultKeepById(vaultKeepId);
        if (keepClone == null)
        {
            throw new Exception($"Invalid Kept Keep id: {vaultKeepId}");

        }
        return keepClone;
    }

    internal string DeleteVaultKeep(string userId, int vaultKeepId)
    {
        KeepClone keepClone = GetVaultKeepById(vaultKeepId);
        if (userId != keepClone.CreatorId)
        {
            throw new Exception("No Permission to delete!");
        }
        _repository.DeleteVaultKeep(vaultKeepId);
        return "deleted";
    }
}

File: client/src/services/VaultKeepsService.js
import { AppState } from "../AppState.js";
import { Keep } from "../models/Keep.js";
import { KeepClone } from "../models/KeepClone.js";
import { logger } from "../utils/Logger.js";
import { api } from "./AxiosService.js";

class VaultKeepsService {

    async createVaultKeep(vaultKeepData) {
        const response = await api.post('api/vaultKeeps', vaultKeepData)
        logger.log('created vaultKeep', response.data)
        const newVaultKeep = new KeepClone(response.data)

        // if (AppState.activeAccount?.id == newVaultKeep.creatorId) {
        //     AppState.vaultKeeps.push(newVaultKeep)
        // }
        AppState.activeKeep.kept++
        return newVaultKeep
    }

    async getVaultKeepsByVaultId(vaultId) {
        AppState.vaultKeeps = []
        const response = await api.get(`api/vaults/${vaultId}/keeps`);
        logger.log('vaults keeps had', response.data);
        const keeps = response.data.map(pojo => new KeepClone(pojo));
        AppState.vaultKeeps = keeps

    }


    // async getVaultKeepById(vaultKeepId) {
    //     const response = await api.get(`api/vaultKeeps/${vaultKeepId}`)
    //     logger.log('GOT VAULT KEEP', response.data)
    //     // AppState.activeVaultKeep = vaultKeepId

    // }
    async getVaultKeepById(vaultKeepId) {
        const res = await api.get(`api/vaultKeeps/${vaultKeepId}`)
        logger.log('vault keep had', res.data)
        AppState.activeVaultKeep = new KeepClone(res.data)

    }


    async deleteVaultKeep(vaultKeepId) {
        const response = await api.delete(`api/vaultKeeps/${vaultKeepId}`)
        logger.log('destroyed vaultKeep', response.data)
    }

}

export const vaultKeepsService = new VaultKeepsService()